### ISSUE TITLE
twisted.names.hosts.searchFileFor returns IPv6 addresses but caller only expects IPv4 addresses

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5098]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5098|
|Type|defect|
|Created|2011-05-09 16:36:42Z|
|Branch|https://github.com/twisted/twisted/tree/ipv6-hosts-5098|

If a hosts file contains an IPv6 address, the default resolver created by Twisted Names' `createResolver` function will not be able to resolve that name.  It will find the IPv6 address and then try to construct an `A` record with it, which is illegal.

Attachments:

* [ipv6-hosts.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/47a/47a78cc892e29d89a7abcdf4feb7afa6cf744ba6/390558dbd8a72612d3d14d6adf35a0d9171c529b.patch) (2974 bytes) - added by lewq on 2011-05-09 16:41:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5098 5098
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_ipv6_hosts_5098 branches/ipv6-hosts-5098
branch_author__lewq__exarkun lewq, exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1304959002000000 1304959002000000
changetime__1309441429000000 1309441429000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="lewq's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|lewq commented|
|-|-|

Traceback demonstrating this problem is:

```
2011-05-09 16:13:46+0000 [-] Unhandled Error
        Traceback (most recent call last):
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/application/app.py", line 392, in startReactor
            self.config, oldstdout, oldstderr, self.profiler, reactor)
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/application/app.py", line 313, in runReactorWithLogging
            reactor.run()
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/internet/base.py", line 1162, in run
            self.mainLoop()
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/internet/base.py", line 1171, in mainLoop
            self.runUntilCurrent()
        --- <exception caught here> ---
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/internet/base.py", line 793, in runUntilCurrent
            call.func(*call.args, **call.kw)
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/internet/tcp.py", line 619, in resolveAddress
            d = self.reactor.resolve(self.addr[0])
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/internet/base.py", line 560, in resolve
            return self.resolver.getHostByName(name, timeout)
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/names/common.py", line 209, in getHostByName
            return self.lookupAllRecords(name, timeout
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/names/resolve.py", line 54, in lookupAllRecords
            d = self.resolvers[0].lookupAllRecords(name, timeout)
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/names/hosts.py", line 73, in lookupAddress
            (dns.RRHeader(name, dns.A, dns.IN, self.ttl, dns.Record_A(res, self.ttl)),), (), ()
          File "/usr/local/lib/python2.6/site-packages/Twisted-11.0.0-py2.6-freebsd-8.1-RELEASE-amd64.egg/twisted/names/dns.py", line 683, in __init__
            address = socket.inet_aton(address)
        socket.error: illegal IP address string passed to inet_aton
```

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [32150]) Branching to 'ipv6-hosts-5098'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [32151]) Apply patch

refs [#5098](https://github.com/twisted/twisted/issues/5098)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli set owner to lewq|
|-|-|

Thank you for working on this lewq!  And thank you for the additional hosts tests.  Without the change to t.n.hosts.py the new tests all pass except for the ignore IPv6 test, which correctly fails.  With the change to t.n.hosts.py, all the new tests pass.

However, the change causes an old test (test_searchFileFor in twisted.names.test.test_names) to fail.  It has an assertion for searching for a host with an IPv6 and expects a result of "::1", but now None is returned instead. That assertion should probably either be deleted or changed.

The fact that `searchFileFor` ignores IPv6 should probably be also be in its docstring.

 --

Another concern I have is whether this change violates Twisted's [compatibility policy](http://twistedmatrix.com/trac/wiki/CompatibilityPolicy).  I don't think the compatibility policy doesn't cover this explicitly, especially since the docstring of the `searchFileFor` method does not specifically say IPv6 is supported, but it does change the returned value.   This is a public function though, and someone's code could conceivably depend on it returning an IPv6 address.

One possibility is to create a new method `searchFileForIPV4` or something like that.  Alternately, perhaps a change can be made in createResolver instead.

Note that I am merely expressing doubt - I do not know enough about twisted.names to evaluate whether this concern is valid, and input from a core developer would probably be useful.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks for the excellent review cyli.  I think your suggestion about making a change to `createResolver` instead is exactly the right thing to do.  Even though the IPv6 behavior of `searchFileFor` isn't documented, the fact that we have a unit test for it is a very strong signal that it should be preserved.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

The branch now teaches `twisted.names.hosts.Resolver` to ignore IPv6 records when computing the result for `lookupAddress`.  It also adds a `lookupIPv6Address` implementation, since it was easy.  And I moved the existing `twisted.names.hosts` tests out of `test_names.py` and into `test_hosts.py`.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ipv6-hosts-5098)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @exarkun|
|-|-|

Looks good, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [32203]) Merge ipv6-hosts-5098

Author: lewq, exarkun
Reviewer: cyli, therve
Fixes: [#5098](https://github.com/twisted/twisted/issues/5098)

Teach `twisted.names.hosts` to ignore IPv6 addresses when looking up IPv4
addresses.  Also implement an IPv6 address lookup method for the hosts
resolver.
